modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/kill/LocalToClusterKillHandlerWrapper.java

```diff
@@ -0,0 +1,149 @@
+
+    /** Maximum time to wait for a remote response. */
+    private static final long RESPONSE_TIMEOUT_MS = TimeUnit.MINUTES.toMillis(5);
+
+    private final OperationKillHandler localHandler;
+    private final CancellableOperationType type;
+    private final String localNodeName;
+    private final LogicalTopologyService logicalTopologyService;
+    private final MessagingService messageService;
+
+    LocalToClusterKillHandlerWrapper(
+            OperationKillHandler localHandler,
+            CancellableOperationType type,
+            String localNodeName,
+            LogicalTopologyService logicalTopologyService,
+            MessagingService messageService
+    ) {
+        assert localHandler.local() : "handler must be local";
+
+        this.localHandler = localHandler;
+        this.type = type;
+        this.localNodeName = localNodeName;
+        this.logicalTopologyService = logicalTopologyService;
+        this.messageService = messageService;
+    }
+
+    @Override
+    public boolean local() {
+        return false;
+    }
+
+    @Override
+    public CancellableOperationType type() {
+        return type;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> cancelAsync(String operationId) {
+        return localHandler.cancelAsync(operationId)
+                .thenCompose((result) -> {
+                    if (Boolean.TRUE.equals(result)) {
+                        return CompletableFuture.completedFuture(Boolean.TRUE);
+                    }
+
+                    CancelOperationRequest request = FACTORY.cancelOperationRequest()
+                            .operationId(operationId)
+                            .typeId(type.id())
+                            .build();
+
+                    return broadcastCancel(request);
+                });
+    }
+
+    private CompletableFuture<Boolean> broadcastCancel(CancelOperationRequest request) {
+        CompletableFuture<Boolean> result = new CompletableFuture<>();
+
+        CompletableFuture<?>[] futures = logicalTopologyService.localLogicalTopology().nodes()
+                .stream()
+                .filter(node -> !node.name().equals(localNodeName))
+                .map(node -> messageService.invoke(node, request, RESPONSE_TIMEOUT_MS)
+                        .thenAccept(msg -> {
+                            CancelOperationResponse response = (CancelOperationResponse) msg;
+                            Throwable remoteErr = response.error();
+
+                            if (remoteErr != null) {
+                                LOG.warn("Remote node returned an error while canceling the operation "
+                                                + "[operationId={}, typeId={}, node={}].", remoteErr,
+                                        request.operationId(), request.typeId(), node.name());
+                            }
+
+                            if (response.error() != null) {
```

Comment:
first, it's better to merge this `if` and previous one. Second, exception thrown from this if will trigger following `whenComplete`, resulting in warning printed to log twice.

Reply:
Thanks. This code was reworked. Now logging is performed only once.